Brought projects to current ver. Android Studio and Gradle

gregko opened the pull request:
Hello!
Thank you for such an elegant and compact, yet powerful and well documented library. I had troubles building it with the old tools, so re-worked projects structure to work with the latest versions of Android Studio and Gradle.  The test examples needed to deal with the new way of handling screen overlay permission under Android 6 and higher, which I added to the test app main activity class.

Also made some small code corrections to remove depreciated or removed calls and classes, e.g. using now NotificationCompat.Builder.

Consider merging it into your base fork, may help others quickly start with your library. I need it for my \@Voice Aloud Reader app to make a floating Play/Pause button and maybe a few other buttons on a floating toolbar.

gregko committed:
Brought projects to current ver. Android Studio and Gradle

Also made some small code corrections to remove depreciated or removed calls and classes, e.g. using now NotificationCompat.Builder

pingpongboss commented on .idea/compiler.xml:
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
this is a contentious subject, but I'd like to remove all of the contents of `.idea` and make it gitignored. This is especially true now that you've helpfully made the project gradle friendly.

pingpongboss commented on StandOutLib/build.gradle:
```diff
@@ -0,0 +1,33 @@
+apply plugin: 'com.android.library'
```
I see you've renamed library/ to StandOutLib/ 

please rename back to library/ to keep the directory structure as before

pingpongboss commented on StandOutLib/build.gradle:
```diff
@@ -0,0 +1,33 @@
+}
+
+dependencies {
+    compile fileTree(dir: 'libs', include: ['*.jar'])
```
no need for this

pingpongboss commented on StandOutLib/build.gradle:
```diff
@@ -0,0 +1,33 @@
+    androidTestCompile('com.android.support.test.espresso:espresso-core:2.2.2', {
+        exclude group: 'com.android.support', module: 'support-annotations'
+    })
+    compile 'com.android.support:appcompat-v7:25.3.1'
```
Does the library need an appcompat dependency?

pingpongboss commented on StandOutLib/src/main/java/wei/mark/standout/StandOutWindow.java:
```diff
@@ -655,9 +656,18 @@ public Notification getPersistentNotification(int id) {
-		Notification notification = new Notification(icon, tickerText, when);
-		notification.setLatestEventInfo(c, contentTitle, contentText,
-				contentIntent);
+		Notification notification = new NotificationCompat.Builder(this)
```
use the application context unless using `this` solves some issue. Same below.

pingpongboss commented on StandOutLib/src/main/java/wei/mark/standout/StandOutWindow.java:
```diff
@@ -655,9 +656,18 @@ public Notification getPersistentNotification(int id) {
+
+//		Notification notification = new Notification(icon, tickerText, when);
+//		notification.setLatestEventInfo(c, contentTitle, contentText,
+//				contentIntent);
```
remove commented lines. Same below

pingpongboss commented on StandOutLib/src/test/java/wei/mark/standout/ExampleUnitTest.java:
```diff
@@ -0,0 +1,17 @@
+public class ExampleUnitTest {
+    @Test
+    public void addition_isCorrect() throws Exception {
+        assertEquals(4, 2 + 2);
```
Please remove or make relevant to the library.

pingpongboss commented on app/build.gradle:
```diff
@@ -0,0 +1,48 @@
+apply plugin: 'com.android.application'
```
I noticed you renamed example/ to app/. Please rename back to example/ to minimize the changes in this PR

pingpongboss commented on app/build.gradle:
```diff
@@ -0,0 +1,48 @@
+    compileSdkVersion 25
+    buildToolsVersion "25.0.3"
+    defaultConfig {
+        applicationId "hyperionics.com.standouttest"
```
rename id to match the package: wei.mark.example

pingpongboss commented on app/build.gradle:
```diff
@@ -0,0 +1,48 @@
+        versionName "1.0"
+        testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
+    }
+    signingConfigs {
```
should not be necessary

pingpongboss commented on app/build.gradle:
```diff
@@ -0,0 +1,48 @@
+}
+
+dependencies {
+    compile fileTree(dir: 'libs', include: ['*.jar'])
```
remove

pingpongboss commented on app/build.gradle:
```diff
@@ -0,0 +1,48 @@
+
+dependencies {
+    compile fileTree(dir: 'libs', include: ['*.jar'])
+    androidTestCompile('com.android.support.test.espresso:espresso-core:2.2.2', {
```
don't see the point in setting up testing for an example app

pingpongboss commented on app/src/main/AndroidManifest.xml:
```diff
@@ -0,0 +1,39 @@
+        <activity android:name="wei.mark.standouttest.MainActivity"
+            android:excludeFromRecents="true"
+            android:theme="@style/Theme.AppCompat.Translucent"
+            >
```
join with line above

pingpongboss commented on app/src/main/AndroidManifest.xml:
```diff
@@ -0,0 +1,39 @@
+<?xml version="1.0" encoding="utf-8"?>
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
+    package="hyperionics.com.standouttest">
```
restore package name back to wei.mark.example. Then you can refer to the service below with `.MostBasicWindow` like before

pingpongboss commented on app/src/main/java/wei/mark/standouttest/MostBasicWindow.java:
```diff
@@ -1,4 +1,4 @@
-package wei.mark.example;
+package wei.mark.standouttest;
```
restore wei.mark.example package names for all these classes

pingpongboss commented on app/src/main/java/wei/mark/standouttest/MainActivity.java:
```diff
@@ -0,0 +1,62 @@
+        StandOutWindow.closeAll(this, WidgetsWindow.class);
+        StandOutWindow.closeAll(this, FloatingFolder.class);
+
+//        StandOutWindow.show(this, MostBasicWindow.class, StandOutWindow.DEFAULT_ID);
```
restore the types of windows displayed to what was previously there

pingpongboss commented on app/src/main/java/wei/mark/standouttest/MultiWindow.java:
```diff
@@ -1,9 +1,10 @@
 import java.util.List;
 import java.util.Locale;
 
+import hyperionics.com.standouttest.R;
```
this can be removed once the package names match each other.

pingpongboss commented on app/src/main/java/wei/mark/standouttest/floatingfolders/FloatingFolder.java:
```diff
@@ -1,4 +1,4 @@
-package wei.mark.floatingfolders;
+package wei.mark.standouttest.floatingfolders;
```
I see you've moved floatingfolders/ into app/.

Restore the old folder hierarchy where example/ and floatingfolders/ are both top level modules. You will have to add an additional build gradle config for floatingfolders/

pingpongboss commented on app/src/main/java/wei/mark/standouttest/MainActivity.java:
```diff
@@ -0,0 +1,62 @@
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_main);
```
do you need to set a content view when the activity isn't meant to display anything?

pingpongboss commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
```
can this file be removed?

pingpongboss commented on app/src/main/res/values/colors.xml:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
```
are these used anywhere? Can they be removed? 

Same question for all other values you've added under res/

pingpongboss commented on app/src/test/java/hyperionics/com/standouttest/ExampleUnitTest.java:
```diff
@@ -0,0 +1,17 @@
+package hyperionics.com.standouttest;
+
```
not relevant, remove

pingpongboss commented on settings.gradle:
```diff
@@ -0,0 +1 @@
+include ':app', ':StandOutLib'
```
you'll want to add floatingfolders here when you make it a module

pingpongboss reviewed (changes requested):
Hey! First of all, thanks for this PR. This will really modernize this library which hasn't been receiving a lot of love in recent years.

I hope you're willing to work with me through this PR. The main theme of my review will be to minimize the number of changes you make. Things like unnecessary package name changes, folder structure changes, gradle configurations, etc, will be reverted so we get a fairly clean PR. Let me know if you have any questions.

gregko commented:
Hi Boss,
Please go ahead and do the changes as you see fit, it’s your library. I’m really too busy to make so many small changes. A few quick answers:

AppCompat dependency – the “new Notification()” simply does not work, one has to use NotificationCompat.Builder, and use AppCompat for compatibility with older versions of Android. I did not change anything else in the library code. However I checked now and see that Notification.Builder was added in API 11, and since the minSdkVersion is set to 15, you may change it to just use Notification.Builder instead and remove appcompat. Oh, another problem, Notification.Builder().setVisibility() requires API 21, unless you use NotificationCompat.Builder()

Using application context – good idea, else we may face memory leaks.

“library” module name – my app uses a lot of library projects, so having one of them named simply “library” would be really confusing for me.

 Units tests and some XML lines you object to – they are created automatically by Android Studio for each new project, feel free to remove them, but I find unit tests and instrumentation test skeletons handy when a need arises to test something.

Wrong package names in some places – sorry about that. Initially I just wanted to make it work asap. Then refactored the package name under Android Studio, and it missed a few places (e.g. the manifest and build.gradle). I’ll correct these in my GitHub fork.

gregko committed:
Fixed some incorrect package names...

Also using application context in creating notifications, to avoid possible memory leaks, restored the test app to show MultiWindow and WidgetsWindow by default (as in the original project), simplified build.gradle, some cosmetic changes.

gregko commented:
In a new commit in my fork I corrected the wrong package names, also changed to use app context in creating notifications, removed unnecessary empty layout for the MainActivity, restored the test app to show MutiWindow and WidgetsWindow by default, as in the original, simplified the app build.gradle and made some other small cosmetic changes. Not sending another pull request at this time, as it seems you'll want to tweak the cosmetics more.

pingpongboss commented:
Thanks gregko.

Sounds like you don't have time to clean this PR up. Let's leave this PR here for now then. If I get any time, I may take over this PR in the future.

Thanks,

gregko committed:
Fixing problem with screen orientation change

Made displayWidth and displayHeight static, added updateScreenSize(context) method, to be called from StandOutWindow service upon configuration change.

gregko committed:
Fix orientation change issue, added ACTION_TOGGLE_VIS

ACTION_TOGGLE_VIS and the code to handle it is for toggling visibility from the persisten notification alternative, instead of having additional notification to show the window. To disable hidden notification, slightly modified the code to ignore it, if the overwritten getHiddenNotificationIntent() returns null.

To handle orientation changes, added onConfigurationChanged() call, which calls Window.updateScreenSize().
